pkg/kb_agent/cronjobs/checkrole.go

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright (C) 2022-2024 ApeCloud Co., Ltd
+
+This file is part of KubeBlocks project
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+package cronjobs
+
+import (
+	"context"
+	"time"
+
+	"github.com/apecloud/kubeblocks/pkg/kb_agent/handlers"
+	"github.com/apecloud/kubeblocks/pkg/kb_agent/util"
+)
+
+type CheckRoleJob struct {
+	CommonJob
+	originRole      string
+	FailedCount     int
+	ReportFrequency int
+}
+
+func NewCheckRoleJob(commonJob CommonJob) *CheckRoleJob {
+	return &CheckRoleJob{
+		CommonJob:       commonJob,
+		originRole:      "waitForStart",
+		ReportFrequency: 60,
+	}
+}
+
+func (job *CheckRoleJob) Do() error {
+	ctx1, cancel := context.WithTimeout(context.Background(), time.Duration(job.TimeoutSeconds))
+	defer cancel()
+	resp, err := handlers.Do(ctx1, job.Name, nil)
+
+	if err != nil {
+		logger.Info("executing checkRole error", "error", err.Error())
+		if job.FailedCount%job.ReportFrequency == 0 {
+			logger.Info("role checks failed continuously", "times", job.FailedCount)
+			err = util.SentEventForProbe(context.Background(), resp)
+		}
+		job.FailedCount++
+		return err
+	}
+
+	role, _ := resp["output"].(string)
+	job.FailedCount = 0
+	if job.originRole == role {
```

Comment:
This logic has a pitfall: if a roleChange event is lost, the controller will not receive any subsequent events unless the role changes again.

A modification to address this issue is to move the logic for detecting role changes into the controller. The kb-agent can push the result of each role probe to the controller, which then determines if a role change has occurred. If an event is occasionally lost, it won't matter, as the next role probe result will still be received.

Reply:
This comment does not require changes in this PR, can be added as a TODO.

Reply:
Yes, the role change event operates on a level trigger mechanism, which involves a tradeoff between accuracy and performance. If the kb-agent were to push an event to the controller for every role probe, it could potentially lead to a burst of events, potentially overwhelming the k8s-apiserver and event controller. is it better to enhance the implementation of the kb-agent to guarantee the successful transmission of the role change event?

Reply:
We could have the kb-agent periodically (e.g., every 1 or 5 minutes to avoid impacting performance) send an event to the controller to inform it of the role. This ensures that even if an event is lost, the controller will eventually receive the information, thereby ensuring correctness.